### PR TITLE
Line search

### PR DESCRIPTION
**Pull request by IvarStefansson**
## Proposed changes

This PR adds functionality for a line search to help convergence of multiphysics problems with fracture deformation.

Some of the functionality could be tested. To be discussed with reviewers. Same goes for naming and placement of code.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tutorials/solution_strategies.ipynb [hunk 1]
```diff
@@ -0,0 +1,224 @@
+{
```
Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <p>The first sentence is aspirational at the moment. Should we add a comment that, for now, this includes only a line search method? Or that the tutorial is under construction?</p><p> </p><p>Can we also add (somewhere) a short description of how to use this as a 'standard' global line search?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <p>For pedagogical reasons, I suggest to split the model setup and the numerics setup into at least two different cells. That should also make it easier to differ between the core of the tutorial (numerics) and what should be considered boilerplate in this context (model). I found myself wanting to ask for much more documentation on the latter part, but then realized that is not the point here, though I did not fully manage to constrain myself.</p><p> </p><p>Put another way: I think it should be clear exactly what is needed to invoke a line search method for contact mechanics problems, and ideally it should be indicated what is needed to make this work as a global line search. Right now, some guesswork is needed to identify the relevant parts of the code. The impressive thing is, if I guess correctly, remarkably few lines are needed.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #74.            sd = subdomains[0]
</pre><p>Why not a loop over all subdomains? I guess from L114 that you know that the domain has only one fracture, and that you further know that this particular function will only be called on fracture subdomains. Neither of those will be clear to a reader of the tutorial, so I suggest adding some comments.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #78.            ) / self.params.get("characteristic_displacement_scaling", 1.0)
</pre><p>Is this different from the characteristic displacement given in the solid material (L108)? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #88.        pp.models.solution_strategy.ContactIndicators,
</pre><p>Here I suggest a short remark that this is the mixin that makes the model and the line search talk, while the rest is standard.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <p>Would it work to pass two different parameter objects, one for the model and one for numerics? I know we don't have good practices for this, but perhaps now is a good time to get started.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <p>These comments are fair, but also warrant a remark at the beginning about this being work in progress.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

**Review by keileg — COMMENTED** [review]
Partial review; I made to the point where the line search got technical. TBC.

**Review by IvarStefansson — COMMENTED** [review]
I think I've addressed most comments. Please let me know what you think, @keileg.

File: tutorials/solution_strategies.ipynb (hunk 1 above)
Comment (IvarStefansson) [reply tutorials/solution_strategies.ipynb:1]: <p>I'd love to do that if I were confident about which parameters to put where. Shall we make an issue?</p>

Comment (IvarStefansson) [reply tutorials/solution_strategies.ipynb:1]: <p>You're right, that would have been better. Outdated now, though.</p>

Comment (IvarStefansson) [reply tutorials/solution_strategies.ipynb:1]: Yes. But it was a bug (from the old version of the code).

Comment (IvarStefansson) [reply tutorials/solution_strategies.ipynb:1]: Done.

File: tutorials/solution_strategies.ipynb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
```
Comment (keileg) [inline tutorials/solution_strategies.ipynb:1]: <p>Thanks for the cleanup, readability is much improved. I see you still use the same parameter dictionary for physics and numerics, but I suggest we discuss this in more general terms in a different context. So, I have no requests for further changes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1208/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: tutorials/solution_strategies.ipynb (hunk 1 above)
Comment (keileg) [reply tutorials/solution_strategies.ipynb:1]: Good idea.

File: src/porepy/numerics/nonlinear/line_search.py [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+
+class LineSearchNewtonSolver(pp.NewtonSolver):
+    """Class for relaxing a nonlinear iteration update using a line search.
+
+    This class extends the iteration method to include a line search and implements
+    a line search based on the residual norm.
+
+    """
+
+    def iteration(self, model) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/nonlinear/line_search.py:50]: I think you are right, I don't know how to do that. Up to know, we have always worked on the inside of the model, where this question does not pop up.

Feel free to resolve.

File: tutorials/solution_strategies.ipynb (hunk 1 above)
Comment (keileg) [reply tutorials/solution_strategies.ipynb:1]: #1210

**Comment by keileg** [comment]
List of unresolved questions/issues:
1. The need to almost copy equation definitions from the models to the numerics-model interface, [here](https://github.com/pmgbergen/porepy/pull/1208/files#r1708846240) 
2. [absolute and relative tolerances in line search](https://github.com/pmgbergen/porepy/pull/1208/files#r1708873687)

I suggest we take not of these, close and make a new PR.

File: tutorials/solution_strategies.ipynb (hunk 2 above)
Comment (IvarStefansson) [reply tutorials/solution_strategies.ipynb:1]: Yes, that was my thinking as well.

**Comment by IvarStefansson** [comment]
> List of unresolved questions/issues:
> 
> 1. The need to almost copy equation definitions from the models to the numerics-model interface, [here](https://github.com/pmgbergen/porepy/pull/1208/files#r1708846240)
> 2. [absolute and relative tolerances in line search](https://github.com/pmgbergen/porepy/pull/1208/files#r1708873687)
> 
> I suggest we take note of these, close and make a new PR.

**Pull request closed without merging**